osqp_vendor/CMakeLists.txt

```diff
@@ -33,12 +33,12 @@ macro(build_osqp)
     DESTINATION
       ${CMAKE_INSTALL_PREFIX}
     USE_SOURCE_PERMISSIONS
-    PATTERN config.h EXCLUDE
   )
 endmacro()
 
 build_osqp()
 
+ament_export_include_directories(include/osqp)
```

Comment:
Please revert this change, it'll break any downstream package that currently depend on `osqp_vendor`, hierarchy headers are better than flat ones, there's fewer chances of collision.

Reply:
I agree, but this is what `osqp` defaults to. Hence not sure how one would resolve this for `osqp_eigen_vendor`

Reply:
Still, downstream ROS packages expect the `osqp` prefix. For `osqp_eigen_vendor`, I'd patch it so that it includes `osqp` via `osqp/osqp.h`

Reply:
does that mean you would patch it in [osqp-eigen](https://github.com/robotology/osqp-eigen). Thanks btw, really appreciate your help

Reply:
I think that PRs in [osqp](https://github.com/osqp/osqp) and [osqp-eigen](https://github.com/robotology/osqp-eigen) are required to fix this. In the meantime, I created PR #13 to fix issue #11.

Please feel free to close this PR. I will take the learnings to create `osqp_eigen_vendor` in a separate repository.

Reply:
Hi, speaking as part of the mantainer team of osqp-eigen. : ) 

There is any reason why you do not just call `ament_export_targets`, as document in ament docs: https://docs.ros.org/en/foxy/How-To-Guides/Ament-CMake-Documentation.html#building-a-library ? 

CMake's imported target are explicitly designed to avoid the to manually handle dependencies via `<pkg>_INCLUDE_DIRS` and `<pkg>_LIBRARIES` variables.

Reply:
so the reason I called `ament_export_include_directories` is that `osqp_vendor` doesn't seem to export the includes, for a reason I do not understand. For example, this very simple project won't compile using `osqp_vendor`

CMakeLists.txt
```cmake
cmake_minimum_required(VERSION 3.5)
project(test)

find_package(ament_cmake REQUIRED)
find_package(osqp_vendor REQUIRED)

add_executable(node main.cpp)

# # doesn't work (preferred)
# ament_target_dependencies(node osqp_vendor)

# works
target_link_libraries(node osqp::osqp)

ament_package()

```
main.cpp
```cpp
#include <osqp.h>

int main(int argc, char** argv) {

    return 0;
}
```
If `ament_export_include_directories` is added, then `ament_target_dependencies(node osqp_vendor)` works.

Reply:
Why can't you simply have `find_package(osqp REQUIRED)` instead of `find_package(osqp_vendor REQUIRED)` ? I tought that the whole point of `<pkg>_vendor` packages is just to provide the `<pkg>`  in a colcon/bloom friendly way, nothing more.

Reply:
you can also do `find_package(osqp REQUIRED)` and link via `target_link_libraries(node osqp::osqp)`. But in this case, ament doesn't work at all.

```cmake
cmake_minimum_required(VERSION 3.5)
project(test)

find_package(ament_cmake REQUIRED)
find_package(osqp REQUIRED) 
# works if ament_export_include_directories added
# # find_package(osqp_vendor REQUIRED) (works, if ament_export_include_directories)

add_executable(node main.cpp)

# doesn't work
ament_target_dependencies(node osqp)  

# works if ament_export_include_directories added
# # ament_target_dependencies(node osqp)

ament_package()
```

Reply:
Ah sorry, I got the point now. I think in this part of the code what you should add is `ament_export_targets(osqp::osqp)`, but I am not sure if that would work. Probably you need to first create a new empty interface target, link that one to `osqp::osqp` and then export that one.

Reply:
Something like:
~~~
add_library(osqp_vendor INTERFACE)
target_link_library(osqp_vendor INTERFACE osqp::osqp)
install(
  TARGETS osqp_vendor 
  EXPORT osqp_vendorTargets)
ament_export_targets(osqp_vendorTargets)
~~~

Not sure if this is the preferred way, but probably the best place to discuss this is to open an issue in ament_cmake I guess.

Reply:
thanks for all the feedback, I'll try my best to dig down to the root cause, but I feel my knowledge of what is happening is too limited at this point